### PR TITLE
implement the external timestamp client

### PR DESCRIPTION
**Pull request by YangKeao**
Signed-off-by: YangKeao <yangkeao@chunibyo.icu>

This PR also upgrades the integration test to 1.19, as TiDB cannot compile under 1.18.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -58,3 +58,5 @@ require (
 	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
 	stathat.com/c/consistent v1.0.0 // indirect
 )
+
+replace github.com/tikv/pd/client v0.0.0-20220725055910-7187a7ab72db => github.com/lhy1024/pd/client v0.0.0-20221026120600-7cf30bf5ce7d
```
Comment (lhy1024) [inline go.mod:62]: plz update kvproto

File: internal/mockstore/mocktikv/pd.go [hunk 2]
```diff
@@ -116,6 +120,32 @@ func (c *pdClient) GetLocalTSAsync(ctx context.Context, dcLocation string) pd.TS
...
+		return err
+	}
+
+	currentTSO := oracle.ComposeTS(p, l)
+	if newTimestamp > currentTSO {
+		return errors.New("external timestamp is greater than global tso")
+	}
+	for {
+		externalTimestamp := c.externalTimestamp.Load()
+		if externalTimestamp > newTimestamp {
```
Comment (lhy1024) [inline internal/mockstore/mocktikv/pd.go:135]: we also may skip equal case, which is unnecessary to update.

File: oracle/oracles/local_external_timestamp.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	if err != nil {
+		return err
+	}
+
+	if newTimestamp > currentTSO {
+		return errors.New("external timestamp is greater than global tso")
+	}
+	for {
+		externalTimestamp := l.externalTimestamp.Load()
+		if externalTimestamp > newTimestamp {
```
Comment (lhy1024) [inline oracle/oracles/local_external_timestamp.go:40]: ditto

File: internal/mockstore/mocktikv/pd.go (hunk 2 above)
Comment (YangKeao) [reply internal/mockstore/mocktikv/pd.go:135]: Fixed.

File: oracle/oracles/local.go [hunk 4]
```diff
@@ -51,6 +51,8 @@ type localOracle struct {
 	hook            *struct {
 		currentTime time.Time
 	}
+
+	localExternalTimestamp
```
Comment (lhy1024) [inline oracle/oracles/local.go:55]: Do we need to set local external ts?

Comment (YangKeao) [reply oracle/oracles/local.go:55]: The empty value should be valid :thinking: , it only contains an atomic 0.

**Review by lhy1024 — APPROVED** [review]

File: integration_tests/go.mod [hunk 5]
```diff
@@ -69,21 +76,26 @@ require (
...
+	google.golang.org/protobuf v1.28.1 // indirect
 	gopkg.in/natefinch/lumberjack.v2 v2.0.0 // indirect
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
 
 replace github.com/tikv/client-go/v2 => ../
+
+replace (
+	github.com/pingcap/tidb v1.1.0-beta.0.20220902042024-0482b2e83ed2 => github.com/YangKeao/tidb v1.1.0-beta.0.20221027084838-c574366d0e5a
+	github.com/pingcap/tidb/parser v0.0.0-20220724090709-5484002f1963 => github.com/YangKeao/tidb/parser v0.0.0-20221027084838-c574366d0e5a
```
Comment (YangKeao) [inline integration_tests/go.mod:100]: They will be removed after merging the https://github.com/pingcap/tidb/pull/38689

**Comment by sticnarf** [comment]
The version policy of client-go is supporting the oldest version under official maintanence. I'm afraid we should stick to 1.18 at present.

The uber atomic can be an alternative right now.

**Comment by YangKeao** [comment]
> The version policy of client-go is supporting the oldest version under official maintanence. I'm afraid we should stick to 1.18 at present.
> 
> The uber atomic can be an alternative right now.

Thanks for the information. I could use uber atomic for implementation and unit test. For integration test, I have to upgrade it to 1.19, as TiDB cannot compile under 1.18 now :thinking: .

**Comment by YangKeao** [comment]
> The version policy of client-go is supporting the oldest version under official maintanence. I'm afraid we should stick to 1.18 at present.
> 
> The uber atomic can be an alternative right now.

Fixed. PTAL @sticnarf

**Comment by sticnarf** [comment]
> Thanks for the information. I could use uber atomic for implementation and unit test. For integration test, I have to upgrade it to 1.19, as TiDB cannot compile under 1.18 now  .

Yep, it's okay. Only the main lib needs to compile on 1.18.

**Pull request merged**